sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -0,0 +1,361 @@
+  std::string extStr(extSize, '\0');
+  ret_err = clGetPlatformInfo(curPlatform, CL_PLATFORM_EXTENSIONS, extSize,
+                              &extStr.front(), NULL);
+
+  if (ret_err != CL_SUCCESS ||
+      extStr.find("cl_khr_il_program") == std::string::npos) {
+    if (res_program != nullptr)
+      *res_program = nullptr;
+    return PI::cast<pi_result>(CL_INVALID_CONTEXT);
+  }
+
+  using apiFuncT =
+      cl_program(CL_API_CALL *)(cl_context, const void *, size_t, cl_int *);
+  apiFuncT funcPtr =
+      reinterpret_cast<apiFuncT>(clGetExtensionFunctionAddressForPlatform(
+          curPlatform, "clCreateProgramWithILKHR"));
+
+  assert(funcPtr != nullptr);
+  if (res_program != nullptr)
+    *res_program = PI::cast<pi_program>(funcPtr(
+        PI::cast<cl_context>(context), il, length, PI::cast<cl_int *>(&err)));
+
+  return err;
+}
+
+pi_result OCL(piSamplerCreate)(pi_context context,
+                               const pi_sampler_properties *sampler_properties,
+                               pi_sampler *result_sampler) {
+  // Initialize properties according to OpenCL 2.1 spec.
+  pi_result error_code;
+  pi_bool normalizedCoords = PI_TRUE;
+  pi_sampler_addressing_mode addressingMode = PI_SAMPLER_ADDRESSING_MODE_CLAMP;
+  pi_sampler_filter_mode filterMode = PI_SAMPLER_FILTER_MODE_NEAREST;
+
+  // Unpack sampler properties
+  for (std::size_t i = 0; sampler_properties && sampler_properties[i] != 0;
+       ++i) {
+    if (sampler_properties[i] == PI_SAMPLER_INFO_NORMALIZED_COORDS) {
+      normalizedCoords = static_cast<pi_bool>(sampler_properties[++i]);
+    } else if (sampler_properties[i] == PI_SAMPLER_INFO_ADDRESSING_MODE) {
+      addressingMode =
+          static_cast<pi_sampler_addressing_mode>(sampler_properties[++i]);
+    } else if (sampler_properties[i] == PI_SAMPLER_INFO_FILTER_MODE) {
+      filterMode = static_cast<pi_sampler_filter_mode>(sampler_properties[++i]);
+    } else {
+      assert(false && "Cannot recognize sampler property");
+    }
+  }
+
+  // Always call OpenCL 1.0 API
+  *result_sampler = PI::cast<pi_sampler>(clCreateSampler(
+      PI::cast<cl_context>(context), normalizedCoords, addressingMode,
+      filterMode, PI::cast<cl_int *>(&error_code)));
+  return error_code;
+}
+
+pi_result OCL(piextGetDeviceFunctionPointer)(pi_device device,
+                                             pi_program program,
+                                             const char *func_name,
+                                             pi_uint64 *function_pointer_ret) {
+  pi_platform platform;
+  cl_int ret_err =
+      clGetDeviceInfo(PI::cast<cl_device_id>(device), PI_DEVICE_INFO_PLATFORM,
+                      sizeof(platform), &platform, nullptr);
+
+  if (ret_err != CL_SUCCESS) {
+    return PI::cast<pi_result>(ret_err);
+  }
+
+  using FuncT =
+      cl_int(CL_API_CALL *)(cl_device_id, cl_program, const char *, cl_ulong *);
+
+  // TODO: add check that device supports corresponding extension
```

Comment:
Is it necessary to add this check on every call to this function? Seems it adds unnecessary overhead/repetition - even if its not much.
Maybe its better to have a PI interface to query extensions, so the SYCL runtime can query once and decide if Function Pointers are possible or not.

Reply:
@smaslov-intel